target_snowflake/sinks.py

```diff
@@ -1,44 +1,140 @@
+from target_snowflake.stages import NamedStage
 
-class SnowflakeSink(CSVSink):
+import csv
+
+
+class SnowflakeSink(SQLSink):
     """Snowflake target sink class."""
 
+    connector_class = SnowflakeConnector
+    stage_class = NamedStage
+
     def __init__(
         self,
-        target: Target,
+        target: SQLTarget,
         stream_name: str,
         schema: Dict,
         key_properties: Optional[List[str]],
     ) -> None:
         super().__init__(target, stream_name, schema, key_properties)
+        self.prepare_sink()
+
+    @property
+    def table_name(self) -> str:
+        return self.stream_name.split("-")[-1].upper()
+
+    @property
+    def schema_name(self) -> Optional[str]:
+        """Return the target schema name."""
+        return self.config["snowflake"]["schema"].upper()
+
+    @property
+    def database_name(self) -> Optional[str]:
+        """Return the target DB name."""
+        return self.config["snowflake"]["database"].upper()
 
-        self.connection = target.connect()
-        self._has_migrated = False
-        self.migrator = SnowflakeSchemaMigrator(sink=self)
-        self.table_schema = target.table_schema
+    def prepare_sink(self) -> None:
+        self.connector.connection.execute(
+            'CREATE SCHEMA IF NOT EXISTS "{self.schema_name}"'
+        )
+        self.connector.prepare_table(
+            self.full_table_name, schema=self.schema, primary_keys=self.key_properties
+        )
+        self.stage = self.stage_class(self)
 
     @property
     def max_size(self):
         return self.config["batch_size_rows"]
 
-    def start_batch(self, context: dict) -> None:
-        # TODO: perhaps Sync should have a callback hook at the beginning of execution?
-        if not self._has_migrated:
-            self.migrator.sync_table_schema()
-            self._has_migrated = True
-        super().start_batch(context)
+    def _write_csv(self, filepath: Path, records: Iterable[dict]) -> None:
+        """Write a CSV file."""
+        with open(filepath, "wt") as fp:
+            writer = csv.writer(fp, delimiter=",")
+            for i, record in enumerate(records, start=1):
+                if i == 1:
+                    writer.writerow(record.keys())
+
+                writer.writerow(record.values())
+        self.logger.info(f"Completed writing {i} records to file...")
+
+    def _upload_csv_to_table_stage(self, file_path: str, full_table_name: str) -> None:
+        """Upload data to the table's built-in stage."""
+        self.connector.connection.execute(
+            f"PUT file://{file_path} @%{full_table_name};"
+        )
+
+    def _load_csv_files_from_table_stage(self, full_table_name: str) -> None:
+        """Loads all files from the table's built-in stage."""
+        self.connector.connection.execute(
+            f"COPY INTO {full_table_name} "
+            f"FILE_FORMAT = (TYPE = CSV FIELD_DELIMITER = '\\t' SKIP_HEADER = 1);"
```

Comment:
formats can be a bit tricky, esp with null handling (as I [learned the hard way](https://github.com/datamill-co/target-snowflake/pull/31)).

I'll test these and see if there's any issues

Reply:
Also if the stream has primary keys we'll need to do the more complicated MERGE operation: 
https://github.com/transferwise/pipelinewise-target-snowflake/blob/master/target_snowflake/file_formats/csv.py#L37

Reply:
Yeah - The MERGE operation I admittedly have not tackled yet. In theory, there's a placeholder method `SQLConnector.merge_upsert_from_table(target_table_name: str, from_table_name: str, join_keys: List[str])` created for this purpose but I haven't fully tested or implemented a solution with that method yet.